src/java/org/apache/cassandra/journal/Journal.java

```diff
@@ -112,31 +116,92 @@ public class Journal<K, V> implements Shutdownable
     private final WaitQueue segmentPrepared = newWaitQueue();
     private final WaitQueue allocatorThreadWaitQueue = newWaitQueue();
     private final BooleanSupplier allocatorThreadWaitCondition = () -> (availableSegment == null);
+    private final FlusherCallbacks flusherCallbacks;
 
     SequentialExecutorPlus closer;
     //private final Set<Descriptor> invalidations = Collections.newSetFromMap(new ConcurrentHashMap<>());
 
+    private class FlusherCallbacks implements Flusher.Callbacks
+    {
+        private final Queue<WaitingFor> waitingFor = new ConcurrentLinkedQueue<>();
+
+        @Override
+        public void onFlush(long segment, int position)
+        {
+            Iterator<WaitingFor> iter = waitingFor.iterator();
```

Comment:
Can't this introduce blocking, where a waiting thread is stuck behind a later write? Probably we should drain and sort the queue. It would be nice also to use something better than a `ConcurrentLinkedQueue` (but not critical)

Reply:
Switched to MPSC queue, and added an array for draining. Will do a bigger improvement that will introduce callbacks native to Journal in a follow-up.